core/data/core.joke

```diff
@@ -4189,3 +4196,190 @@
+  [multifn]
+  (throw (ex-info "method preference not yet supported by joker.core" {})))
+
+(defmacro defmulti
+  "Creates a new multimethod with the associated dispatch function.
+  The docstring and attr-map are optional.
+
+  Options are key-value pairs and may be one of:
+
+  :default
+
+  The default dispatch value, defaults to :default
+
+  :hierarchy (UNSUPPORTED)
+
+  The value used for hierarchical dispatch (e.g. ::square is-a ::shape)
+
+  Hierarchies are type-like relationships that do not depend upon type
+  inheritance. By default Clojure's multimethods dispatch off of a
+  global hierarchy map.  However, a hierarchy relationship can be
+  created with the derive function used to augment the root ancestor
+  created with make-hierarchy.
+
+  Multimethods expect the value of the hierarchy option to be supplied as
+  a reference type e.g. a var (i.e. via the Var-quote dispatch macro #'
+  or the var special form)."
+  {:arglists '([name docstring? attr-map? dispatch-fn & options])
+   :added "1.0"}
+  [mm-name & options]
+  (let [docstring   (if (string? (first options))
+                      (first options)
+                      nil)
+        options     (if (string? (first options))
+                      (next options)
+                      options)
+        m           (if (map? (first options))
+                      (first options)
+                      {})
+        options     (if (map? (first options))
+                      (next options)
+                      options)
+        dispatch-fn (first options)
+        options     (next options)
+        m           (if docstring
+                      (assoc m :doc docstring)
+                      m)
+        m           (if (meta mm-name)
+                      (conj (meta mm-name) m)
+                      m)
+        mm-name (with-meta mm-name m)]
+    (when (= (count options) 1)
+      (throw (ex-info "The syntax for defmulti has changed. Example: (defmulti name dispatch-fn :default dispatch-value)" {}))) ; was: (throw (Exception. ...))
+    (let [options   (apply hash-map options)
+          default   (get options :default :default)
+          hierarchy (get options :hierarchy nil)]
+      (check-valid-options options :default :hierarchy)
+      `(let [v# (def ~mm-name)]
+         (when-not (and (bound? v#) ; was: (and (.hasRoot v#) (instance? clojure.lang.MultiFn (deref v#)))
+                        (fn? (deref v#))
+                        (:method-table (meta (deref v#))))
+           (let [fndef# (apply new-multifn (list (var ~mm-name) ~dispatch-fn ~default ~hierarchy))]
+                 (def ~mm-name fndef#)))))))
+
+(defmacro defmethod
+  "Creates and installs a new method of multimethod associated with dispatch-value. "
+  {:added "1.0"}
+  [multifn dispatch-val & fn-tail]
+  `(multifn-swap-method-table-vals! ~multifn assoc ~dispatch-val (fn ~@fn-tail)))
+
+(defn remove-all-methods
+  "Removes all of the methods of multimethod."
+  {:added "1.2"
```

Comment:
`:added` should be `1.0` everywhere, as it refers to the version of Joker, not Clojure. Hopefully one day there will be version 1.0...

Reply:
Thanks for the thorough review! Mainly regarding inlining, note that I used Clojure's own `defmulti` code as a template for this effort, and kept the separation of public-function definitions from private-function implementations I saw there (and later saw that ClojureScript copied as well).

I don't mind combining them as suggested, but first want to make sure you're okay with moving the code base further afield from upstream.

(If `defmulti` was a rapidly moving target, i.e. a hotbed of innovation, I'd be more concerned about doing that!)